src/libraries/System.Text.Json/source_generation/src/JsonSourceGeneratorHelper.Generate.cs

```diff
@@ -0,0 +1,672 @@
+                        : "getter: null";
+
+                    string setterNamedArg;
+                    if (memberMetadata.HasSetter)
+                    {
+                        string propMutation = declaringTypeIsValueType
+                            ? @$"{{ Unsafe.Unbox<{declaringTypeCompilableName}>(obj).{clrPropertyName} = value; }}"
+                            : $@"{{ (({declaringTypeCompilableName})obj).{clrPropertyName} = value; }}";
+
+                        setterNamedArg = $"setter: static (obj, value) => {propMutation}";
+                    }
+                    else
+                    {
+                        setterNamedArg = "setter: null";
+                    }
+
+                    JsonIgnoreCondition? ignoreCondition = memberMetadata.IgnoreCondition;
+                    string ignoreConditionNamedArg = ignoreCondition.HasValue
+                        ? $"ignoreCondition: JsonIgnoreCondition.{ignoreCondition.Value}"
+                        : "ignoreCondition: default";
+
+                    string encodedNameArg;
+                    if (!ContainsNonAscii(clrPropertyName))
+                    {
+                        byte[] name = Encoding.UTF8.GetBytes(memberMetadata.JsonPropertyName ?? clrPropertyName);
+                        string nameBytes = string.Join(", ", name.Select(b => $"{b}"));
+                        encodedNameArg = "encodedName: new byte[] {" + nameBytes + "}";
+                    }
+                    else
+                    {
+                        encodedNameArg = "encodedName: default";
+                    }
+
+                    string converterNamedArg = memberMetadata.ConverterInstantiationLogic == null
+                        ? "converter: null"
+                        : $"converter: {memberMetadata.ConverterInstantiationLogic}";
+
+                    string memberTypeCompilableName = memberTypeMetadata.CompilableName;
+
+                    sb.Append($@"
+            {PropVarName}[{i}] = {JsonMetadataServicesClassName}.CreatePropertyInfo<{memberTypeCompilableName}>(
+                options,
+                isProperty: {memberMetadata.IsProperty.ToString().ToLowerInvariant()},
+                declaringType: typeof({memberMetadata.DeclaringTypeCompilableName}),
+                {typeTypeInfoNamedArg},
+                {converterNamedArg},
+                {getterNamedArg},
+                {setterNamedArg},
+                {ignoreConditionNamedArg},
+                numberHandling: {GetNumberHandlingAsStr(memberMetadata.NumberHandling)},
+                propertyName: ""{clrPropertyName}"",
+                {jsonPropertyNameNamedArg},
+                {encodedNameArg});
+            ");
+                }
+            }
+
+            sb.Append(@$"
+            return {PropVarName};
+        }}");
+
+            return sb.ToString();
+
+            static bool ContainsNonAscii(string str)
```

Comment:
@GrabYourPitchforks PTAL at encoding logic here for property names. I believe single quotes and backslashes are the 2 common escaping cases which fall into Ascii range.

Reply:
I've taken a note to verify the semantics here (https://github.com/dotnet/runtimelab/projects/1#card-59161310). For now, I've added logic to signal strings that contain these characters as "needs escaping at runtime"